src/hamiltonian.jl

```diff
@@ -42,16 +42,19 @@ end
 ∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) = copy(r)
 ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) =
     h.metric.M⁻¹ .* r
-∂H∂r(h::Hamiltonian{<:DenseEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) =
-    h.metric.M⁻¹ * r
+function ∂H∂r(h::Hamiltonian{<:DenseEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat)
+    out = similar(r) # Make sure the output of this function is of the same type as r
+    mul!(out, h.metric.M⁻¹, r)
+    out
+end
```

Comment:
I'm a bit uncertain about this change as it "complicates" code to mainly just stay compatible with ComponentArrays.jl, and thus I'd be more in favour of just making it an extension instead, I think :confused: Then in the extension, we just overload whatever we need to be compatible.

Also, will this code break if, say, `h.metric.M⁻¹` has eltype `Float64` but `r` has eltype `Float32`, rather than just promoting, as is current behavior?

Reply:
```Julia
p1 = ComponentArray(m=one(Float32), s = one(Float32))
r = similar(p1)
M = diagm(randn(Float64, 2))
mul!(r, M, p1)
```

This works on my machine, and returns `r` as a component array of eltype `Float32` as expected.